wxMSWHeaderCtrl fixes

MaartenBent opened the pull request:
Fix using `wxHD_BITMAP_ON_RIGHT`.
Possible fix for [#18245](https://trac.wxwidgets.org/ticket/18245).
Show more columns in widgets sample, redesign page.

MaartenBent committed:
Fix applying header style of wxMSWHeaderCtrl

Call SetWindowStyleFlag when creating the control. Apply wxHD_BITMAP_ON_RIGHT
style. Get rid of the helper function.

MaartenBent committed:
Add more columns to Header control in widgets sample

Show column settings horizontally.
Fix binding to header control events after it is recreated.

vadz commented on src/msw/headerctrl.cpp:
```diff
@@ -1136,16 +1142,16 @@ void wxHeaderCtrl::SetWindowStyleFlag(long style)
-        return style | wxHD_ALLOW_REORDER;
+    flags &= ~wxHD_ALLOW_REORDER;
+    if (HasFlag(wxHD_ALLOW_REORDER))
+        flags |= wxHD_ALLOW_REORDER;
```
Very minor, but I'd write this (and the lines below) as
```
if ( HasFlag(wxHD_FOO) )
    flags |= wxHD_FOO;
else
    flags &= ~wxHD_FOO;
```
not because of any efficiency considerations, but just because it seems much more clear what this does to me.

vadz commented on samples/widgets/headerctrl.cpp:
```diff
@@ -200,19 +200,20 @@ void HeaderCtrlWidgetsPage::RecreateWidget()
+                                 GetColumnStyleFlags(i));
+        if ( m_colSettings[i].chkWithBitmap->IsChecked() )
+        {
+            wxArtID artID = i % 4 == 0 ? wxART_ERROR : (i % 4 == 1 ? wxART_QUESTION : (i % 4 == 2 ? wxART_WARNING : wxART_INFORMATION));
```
```suggestion
            wxArtID artID;
            switch ( i % 4 )
            {
                case 0: artID = wxART_ERROR; break;
                case 1: artID = wxART_QUESTION; break;
                case 2: artID = wxART_WARNING; break;
                case 3: artID = wxART_INFORMATION; break;
            }
```
or maybe
```
const wxArtID icons[] = { wxART_ERROR, wxART_QUESTION, wxART_WARNING, wxART_INFORMATION };
wxArtID artID = icons[i % WXSIZEOF(icons)];
```

Both of those seem more readable to me.

PBfordev replied:
I, being in the "never use literals in the code" camp, would also suggest replacing the literal `4` everywhere, including  `m_colSettings` declaration, with a constant or sizeof expression.

MaartenBent replied:
Removed literals, only one define remaining for the number of columns. This makes it easy to quickly adjust the sample to show more or less columns.

vadz reviewed (commented):
Thanks, we should definitely apply this if it fixes the (user-visible) issue, but I have 2 questions:

1. Where is `UpdateHeader()` called from? This doesn't happen under MSW 7 and I didn't have time to debug it under 10 myself yet, so if you already know, could you please let me know? If not, I'll just check it myself later, of course. And the reason for asking is that I wonder if we can avoid calling it in the first place during resizing, rather than returning from it if it happens.
2. Under MSW 7 when I run the widgets sample I don't see the resize cursor when hovering over the column separators, even though I still can resize them. Changing any of the column checkboxes fixes this, e.g. just enabling the bitmap, for example, is enough for the proper resize cursor to be shown. Would you have any idea about this?

a-wi commented:
>    Under MSW 7 when I run the widgets sample I don't see the resize cursor when hovering over the column separators, even though I still can resize them. Changing any of the column checkboxes fixes this, e.g. just enabling the bitmap, for example, is enough for the proper resize cursor to be shown.

MSW 10 is affected too.

a-wi commented:
>     1. Where is `UpdateHeader()` called from?

`wxMSWHeaderCtrl::UpdateHeader()` is called through `wxHeaderCtrlBase::UpdateColumn()` ->`wxHeaderCtrl::DoUpdate()` from `wxDataViewCtrl::OnColumnChange()` and `wxDataViewCtrl::UpdateColWidths()` when `wxDataViewCtrl` is being resized - can be seen in dataview sample.
`wxMSWHeaderCtrl::UpdateHeader()` in **not** called by `wxHeaderCtrl` itself when it is being resized - see widgets sample.
So it seems the root cause of the problem is rather `wxDVC` and not `wxHeaderCtrl`.

vadz commented:
Thanks, I see it now. This definitely looks wrong, but I'm not really sure what's the best way to fix it in wxDVC code.

BTW, we also end up calling `OnColumnChange()` twice, once from `wxDataViewColumn::SetWidth()` and then directly from `wxDataViewHeaderWindow::OnResize()`.

vadz commented:
I'm not sure if it's the best thing to do, but I think #1711 should solve this too in a slightly better way.

Please let me know if you (dis)agree that it's preferable to a535e6a here. TIA!

MaartenBent commented:
I agree that solving it in `wxDataView` is better. I'll remove my fix from this PR and address the other issues.

MaartenBent committed:
Fix reverting to standard cursor in widgets sample

Don't set the cursor to wxSTANDARD_CURSOR, but set it to wxNullCursor so the
default system cursor is used.

MaartenBent commented:
> 2\. Under MSW 7 when I run the widgets sample I don't see the resize cursor when hovering over the column separators, even though I still can resize them. Changing any of the column checkboxes fixes this, e.g. just enabling the bitmap, for example, is enough for the proper resize cursor to be shown. Would you have any idea about this?

This was caused by the widgets sample, it modifies the cursor after page load. I fixed it by using the default cursor instead of `wxSTANDARD_CURSOR`.

vadz commented:
> This was caused by the widgets sample, it modifies the cursor after page load. I fixed it by using the default cursor instead of `wxSTANDARD_CURSOR`.

Ah, thanks, and sorry for not realizing it wasn't a new problem.
